Replace SimpleDateFormat used with ThreadLocal into java.time packages #2250

carryel opened the pull request:
java.text.SimpleDateFormat is now deprecated in Grizzly and ThreadLocal, which was used with SimpleDateFormat, has also been removed.

+ Applied to Cookie utilities.
+ Added HttpDateFormat, which replaces FastHttpDateFormat.
+ Added DateTimeFormatters, which replace SimpleDateFormats.
+ Applied to http/http2/http-server modules.
+ Re-enabled ApacheLogFormatTest, which was ignored in the JDK17 build.

carryel committed:
Issue #2250 "Replace SimpleDateFormat used with ThreadLocal into java.time packages" (https://github.com/eclipse-ee4j/glassfish-grizzly/issues/2250)
+ Applied at Cookie utility.
+ Added HttpDateFormat, which replaces FastHttpDateFormat.
+ Added DateTimeFormatters, which replace SimpleDateFormats.

carryel committed:
Issue #2250 "Replace SimpleDateFormat used with ThreadLocal into java.time packages" (https://github.com/eclipse-ee4j/glassfish-grizzly/issues/2250)
+ Applied to http/http2/http-server modules.
+ Re-enabled ApacheLogFormatTest, which was ignored in the JDK17 build.

avpinchuk commented on modules/http-server/src/main/java/org/glassfish/grizzly/http/server/accesslog/ApacheLogFormat.java:
```diff
@@ -757,9 +774,8 @@ StringBuilder format(StringBuilder builder, Request request, Response response,
-            format.setTimeZone(timeZone);
-            return builder.append(format.format(timeStamp));
+            final DateTimeFormatter format = dateTimeFormatter.withZone(timeZone);
+            return builder.append(format.format(Instant.ofEpochMilli(timeStamp.getTime())));
```
Why not use `timeStamp.toInstant()` instead of `Instant.ofEpochMillis(...)`?

avpinchuk commented on modules/http-server/src/main/java/org/glassfish/grizzly/http/server/util/DateTimeFormatters.java:
```diff
@@ -0,0 +1,56 @@
+
+    private final DateTimeFormatter[] f;
+
+    public DateTimeFormatters() {
```
It's better to make the constructor private.

It is also better to give a more expressive name to the variable `f`, e.g. `formatters`.

carryel committed:
Issue #2250 "Replace SimpleDateFormat used with ThreadLocal into java.time packages" (https://github.com/eclipse-ee4j/glassfish-grizzly/issues/2250)
+ Applied @avpinchuk's feedback:
  - https://github.com/eclipse-ee4j/glassfish-grizzly/pull/2251#discussion_r2410376540
  - https://github.com/eclipse-ee4j/glassfish-grizzly/pull/2251#discussion_r2410352523
  - https://github.com/eclipse-ee4j/glassfish-grizzly/pull/2251#discussion_r2411986839
  - https://github.com/eclipse-ee4j/glassfish-grizzly/pull/2251#discussion_r2414348608

avpinchuk commented on modules/http/src/main/java/org/glassfish/grizzly/http/util/HttpDateFormat.java:
```diff
@@ -0,0 +1,195 @@
+    /**
+     * Get the current date in HTTP format.
+     */
+    public static byte[] getCurrentDateBytes() {
```
IMO this method is dangerous.

Please try something similar (if I haven't got anything wrong)

```java
public static byte[] getCurrentDateBytes() {
        final long now = System.currentTimeMillis();
        final long diff = now - nextGeneration;

        if (diff < 0 || (diff < 5000 && isGeneratingNow.get())) {
            return currentDateBytes;
        }

        while (!isGeneratingNow.compareAndSet(false, true)) {
            Thread.onSpinWait();
        }

        if (now > nextGeneration) {
            currentDateBytes = toCheckedByteArray(new StringBuilder(FORMATTER.format(Instant.ofEpochMilli(now))));
            nextGeneration = now + 1000;
        }

        isGeneratingNow.set(false);

        return currentDateBytes;
    }
```

carryel replied:
Your suggested approach is probably correct. I'll look into it a bit more. I know it's risky, but I allowed it because the logic didn't seem to prioritize consistency. However, I think my approach wasn't a good one.

carryel replied:
I considered several other methods, including ReentrantLock and conditional changes, but your suggestion seemed the most beautiful and efficient, so I adopted it as is.

avpinchuk commented on modules/http/src/main/java/org/glassfish/grizzly/http/util/HttpDateFormat.java:
```diff
@@ -0,0 +1,195 @@
+    /**
+     * Current formatted date as byte[].
+     */
+    private static volatile byte[] currentDateBytes;
```
This may be non-volatile

carryel replied:
I'd like to ask for clarification. Is this because the `Happens-Before` Guarantee/Relationships affect non-volatile variables?
I missed that.

avpinchuk replied:
> Is this because the `Happens-Before` Guarantee/Relationships affect non-volatile variables?

Yes. Let two variables: non-volatile `a` and volatile `b` and two threads 

```
Thread 1:
a = value1
b = value2 // volatile write

Thread 2:
read b // volatile read
read a
```

Then if `read b` = `value2` then `read a` guaranteed to see `value1`.

avpinchuk replied:
@carryel If you have any questions, I'm always happy to help ;)

carryel replied:
I resubmitted the PR, removing the unnecessary volatile declaration.

avpinchuk commented on modules/http/src/main/java/org/glassfish/grizzly/http/util/HttpDateFormat.java:
```diff
@@ -0,0 +1,195 @@
+    /**
+     * Current formatted date.
+     */
+    private static volatile String cachedStringDate;
```
This may be non-volatile too

carryel replied:
I resubmitted the PR, removing the unnecessary volatile declaration.
